### DISCUSSION TITLE
Question about extensive use of RxJS in the codebase

### DISCUSSION THREAD

**Discussion by gioragutt** [jscutlery/semver#390] [General]
Hi maintainers! I've been going over the code today to learn what's being done by the plugin, and I noticed that there's extensive use of RxJS all over the codebase.

Now, don't get me wrong, I LOVE RxJS, I'm not saying it shouldn't be used. But I think that if you don't need it, you're probably better off not using it.

From what I've been reading, I've not seen a case where RxJS is actually needed. Most of the observables are `from(somePromise)` and other one-off values.

So what I'm trying to say is that I'd be happy to hear why RxJS was used the way it was. Purely for education purposes.

**Comment by edbzn** [comment]
Hello, 

Historically before using Nx Devkit we were using the Angular Devkit which is only based on RxJS, at this moment we had no choice to use RxJS. We already knew that using it in our context had no real benefits over Promises (one single emission, no cancel, no retry...), and as you noticed, almost all of our dependencies are Promise-based which implies coercing many sources to an Observable.

When we migrated to the Nx Devkit, we thought it was a great opportunity to refactor using Promises, I realized it was too much work in one single round, I underestimated the time it would take.

One other aspect is getting new contributors, I'm mitigated here because some contributors (including me) would contribute because they love RxJS and they are attracted by open-source projects that use it. On the contrary, some others would never contribute because RxJS is blocking for them.

I think this discussion could be a good place to decide whether we should stay on RxJS or migrate to Promises. What is your opinion?

**Reply by gioragutt** [reply]
First of all, thanks for the reply!

My opinion, as someone who really loves RxJS and has used it extensively, is that if there's no real reason to use it (async environment with many events and whatnot), it's better off to be without it.

Observables have the overhead of being more verbose (for simple cases) and being more difficult to debug.
Due to that, I believe that there's not much reason to use RxJS in this case. This will simplify the codebase and will make it more readable.

At this point, where there's no real reason for RxJS, it'll be easy to migrate to Promises.
And if there'll be a real reason to use Observables, it's probably to keep it contained in something that will probably consume a Promise result anyway.
